Python/import.c

```diff
@@ -1050,17 +1050,78 @@ _imp_create_builtin(PyObject *module, PyObject *spec)
 
 /* Frozen modules */
 
+static bool
+parse_env_var_flag(const char *rawvalue, bool *isset)
+{
+    /* It isn't set or it is set to an empty string. */
+    if (rawvalue == NULL || strlen(rawvalue) == 0) {
+        if (isset != NULL) {
+            *isset = false;
+        }
+        return false;
+    }
+    if (isset != NULL) {
+        *isset = true;
+    }
+    if (strcmp(rawvalue, "0") == 0) {
+        return false;
+    }
+    /* For now we treat all other non-empty strings as true. */
+    return true;
+}
+
+static bool
+is_essential_frozen_module(const char *name)
+{
+    /* These modules are necessary to bootstrap the import system. */
+    if (strcmp(name, "_frozen_importlib") == 0) {
+        return true;
+    }
+    if (strcmp(name, "_frozen_importlib_external") == 0) {
+        return true;
+    }
+    if (strcmp(name, "zipimport") == 0) {
+        return true;
+    }
+    /* This doesn't otherwise have anywhere to find the module.
+       See frozenmain.c. */
+    if (strcmp(name, "__main__") == 0) {
+        return true;
+    }
+    return false;
+}
+
+static bool
+use_frozen(const char *modname)
+{
+    bool isset;
+    /* Note that we don't bother with os.environ. */
+    bool use = parse_env_var_flag(getenv("_PYTHONTESTFROZENMODULES"), &isset);
```

Comment:
You may be thinking of `Py_GETENV()` (in Include/cpython/pydebug.h), which does exactly that.  I considered using it.  At that point I found it was for getting env vars that were passed into the process (rather than ones that were set within the process), which doesn't really apply to how we're using `_PYTHONTESTFROZENMODULES`.  In fact, we never want that env var to be ignored.

FWIW, `PyConfig.use_frozen_modules` is normally set at the commandline via `-X frozen_modules=...` (or with a default).  `_PYTHONTESTFROZENMODULES` is a temporary override used within a handful of tests in our test suite.  I settled on using an environment variable for this because it was the simplest approach.  The only other solution that comes to mind is something like `sys._set_use_frozen_modules_for_test(bool)`, which would directly update `PyConfig`.  However, that felt like extra work (and exposure) for little benefit.  If you think env vars carry too much baggage for this use case I'm okay with taking the `sys` approach (or something else).

Reply:
As to returning -1, I tried that too but found several reasons for the current signature.  However, now that I look at it I don't think `parse_env_var_flag()` pays for itself.  If `bool _Py_ParseEnvVarFlag(const char *)` already existed I would use it.  However, there's no need for this local one-off.  So I've removed it.  Thanks for pointing me over here.

Reply:
> You may be thinking of `Py_GETENV()` (in Include/cpython/pydebug.h), which does exactly that. I considered using it. At that point I found it was for getting env vars that were passed into the process (rather than ones that were set within the process), which doesn't really apply to how we're using `_PYTHONTESTFROZENMODULES`. In fact, we never want that env var to be ignored.

I was thinking of `_Py_GetEnv()` but it's roughly the same thing. (Odd that there's a macro too.)

Since you now set the variable with os.putenv(), the C-level getenv() call in those functions should work.

Another question around this issue: why use an env var at all? Does it ever get passed into a subprocess?

And why do you say you never want this variable ignored? I thought the promise documented for the -E flag was pretty clear.

Reply:
> Another question around this issue: why use an env var at all? Does it ever get passed into a subprocess?

Yeah, it's never passed to a subprocess, which is definitely not typical.  Since you've brought it up, it's making me wonder if it would be better to go with a different solution, even if not as simple, just so no one ends up with the same questions.

FWIW, using an env var was the first simple approach that came to mind and I still think it's the simplest solution.  Like I said, I could have also added something like `sys._set_use_frozen_modules_for_tests()`, but that's more noticeable and therefore more of a potential source of confusion for users (and more of a temptation).

> And why do you say you never want this variable ignored? I thought the promise documented for the -E flag was pretty clear.

The only purpose of the environment variable is to set a temporary override in a few tests in our test suite.  We do not want it ignored because the point is to get consistent test results regardless of how the tests are run.

I suppose we could make it a public alternative to the "-X" option like we do with others (at which point the -E flag should apply), but I'd rather not if we don't need to.  Do you think it would be worth it?

Reply:
Is there no prior art? Are none of the config flags settable (or even readable) from Python code?

Reply:
Good question!  Really, the config is only ever modified in the first phase of runtime initialization.  After that, the config represents what was used to initialize the runtime.  So what I'm doing here is deviating from that.  What I need to be doing is use a new field on `PyInterpreterState` that is initialized to the config value.  The precedent for setting values there is through `sys`.  I'll fix this up.

Reply:
this is done